### ISSUE TITLE
Bug of #10753 still exists on Microfocus (Novell) OES NCP Volumes.

### ISSUE THREAD

**Issue by LeMerP** [notepad-plus-plus/notepad-plus-plus#10992]
### Description of the Issue
The issue with the message "This file has been modified by another program. Do you want to reload it?" still exists on NSS volumes of Microfocus (former Novell) Open Enterprise Server 2018.x.

### Steps to Reproduce the Issue
1. In Windows, create a text file on the NSS volume (\\\\oesserver\nss\test.txt).
2. Open the file and type some text, but do not save the changes.
3. Leave the notepad++ application, but do not close ist.
4. Go back to notepad++ and you will get the reload-message, although the file was not changed by another program.

### Expected Behavior
No reload message should occur.

### Actual Behavior
reload message does occur.

### Debug Information
Notepad++ v8.2   (32-bit)
Build time : Dec 30 2021 - 03:36:49
Path : C:\Program Files (x86)\Notepad++\notepad++.exe
Command Line : "\\\\oesserver\nss\test.txt"
Admin mode : OFF
Local Conf mode : OFF
Cloud Config : OFF
OS Name : Windows 10 Enterprise (64-bit) 
OS Version : 2009
OS Build : 19044.1415
Current ANSI codepage : 1252
Plugins : none

The following debug infos shown in Notepad++ are wrong and should be these ones:
OS Name : Windows 10 Pro
OS Version: 21H2

![Clipboard01](https://user-images.githubusercontent.com/6209900/148023478-86094aa6-5761-489b-b89c-a97f03eabd35.png)

**Comment by xomx** [comment]
Could you please create an empty file with the name 'nppLogNetworkDriveIssue.xml' in the Notepad++ installation directory (at the same place, where is the notepad++.exe file) before you start the Notepad++? This is a logging trigger for some potentially problematic network operations. With some luck, after your issue manifests again, we will obtain a nppLogNetworkDriveIssue.log file (in your %APPDATA%\Notepad++\ directory, or in Notepad++ installed folder for a portable package) with some hints.

(I suspect now the GetFileAttributesEx failing somehow, but without evidence it is just a blind shot.)

**Comment by LeMerP** [comment]
Content of nppLogNetworkDriveIssue.log:
`2022-01-07 08:21:33  \\192.168.56.111\sys\test.txt  in checkFileState(): attributes.ftLastWriteTime (797625344/30933911) < _timeStamp (804519060/30933911)
`

**Comment by LeMerP** [comment]
By the way, I had to change the headline, because the issue affects all NCP volume shares. (https://www.novell.com/documentation/open-enterprise-server-2018/file_ncp_lx/data/btpfqb0.html)

**Comment by xomx** [comment]
Looks like the GetFileAttributesEx does not fail, but it really returns this different timestamp instead.

**Comment by xomx** [comment]
@LeMerP 
Just to be sure - did you also test the older N++ version 8.1.5 (according to the #10753 the last non-problematic one) in your setup?

**Comment by LeMerP** [comment]
Issue also persists in 8.1.5, but not in v7.
Notepad++v7 works like a charm.

**Comment by donho** [comment]
@LeMerP 
That's wired. v8.1.6 (and version later) contains IO API change, but v8.1.5 should have no issue.
However, [v7](https://notepad-plus-plus.org/downloads/v7.0/) was released in 2016, so the change during 5 years could be huge.
If you can find from which version it stop working, I might be able to check code source for your issue.

**Comment by LeMerP** [comment]
@donho 
Version 7.3.3 is the last working one. Issue exists from 7.5.4 onward.

**Comment by donho** [comment]
@LeMerP 
Between v7.3.3 and v7.5.4 there were 4 releases (7.4.2, 7.5.1, 7.5.2 753):
http://download.notepad-plus-plus.org/repository/7.x/

I cannot check if the working version and not working version are not the previous/next version respectively.

**Comment by LeMerP** [comment]
@donho 
ah, sry, I used the 'offical' download site (https://notepad-plus-plus.org/downloads/).
7.5.1 is working as expected. 7.5.2 not.

Debug Info
Notepad++ v7.5.1   (32-bit)
Build time : Aug 29 2017 - 02:35:41
Path : C:\Program Files (x86)\Notepad++\notepad++.exe
Admin mode : OFF
Local Conf mode : OFF
OS : Windows 10 (64-bit)
Plugins : mimeTools.dll NppConverter.dll NppExport.dll

**Comment by donho** [comment]
@LeMerP 
Are you sure the working version is v7.5.1 and not working version is v7.5.2 ?
I have downloaded the source code of both version from:
https://github.com/notepad-plus-plus/notepad-plus-plus/releases?page=5
and compared between 2 versions with WinMerge - I didn't find and IO routine change or timestamp comparison modification between the 2 versions.

**Comment by xomx** [comment]
@donho 
I waited for your opinion, as I did not find anything suspicious either (by simple comparing 7.5.1&2).

**Comment by donho** [comment]
@xomx 
That's why I need the confirmation from @LeMerP  - frome the diff of source code, I don't see any change could cause this different behaviour.

**Comment by LeMerP** [comment]
@donho 
I rechecked again. 7.5.1 is working, 7.5.2 not.

**Comment by donho** [comment]
@LeMerP 
Thank you for rechecking.

@xomx 
Do you have any thought about it?

**Comment by ArkadiuszMichalski** [comment]
The transition to VS2015 / VS2017 did not bring anything new?
https://github.com/notepad-plus-plus/notepad-plus-plus/commits/master?after=90016a8dc0c470882c698c396d1b7e04c38cac0b+1434&branch=master

**Comment by xomx** [comment]
I have found only these suspects:

- PlatformToolset was changed from v120_xp to v140_xp
- the only change in code, which is related to an IO, is in the .\PowerEditor\src\WinControls\FunctionList\functionListPanel.cpp, but how this can be connected I have no idea

**Comment by donho** [comment]
@xomx 
> the only change in code, which is related to an IO, is in the .\PowerEditor\src\WinControls\FunctionList\functionListPanel.cpp, but how this can be connected I have no idea

There's no connected for sure.

@ArkadiuszMichalski @xomx 
Well spotted! I don't see any cause other than the change of toolset and compiler. 

@LeMerP 
I have no more vS2015 + toolset v120_xp to build a binary for testing. Even I had, it couldn't be done to fix this issue. I'll keep this issue opened but unfortunately so far I have no any solution for this problem. Hope you understand.

**Comment by LeMerP** [comment]
Thanks for your time.

@donho 
With that information, Microfocus should talk to Microsoft to solve this issue. I will contact Microfocus support and hope they will get a solution.

I will come back to you as soon as I got more information.

**Comment by donho** [comment]
@LeMerP 
> With that information, Microfocus should talk to Microsoft to solve this issue. I will contact Microfocus support and hope they will get a solution.

Sounds great! That will be the best solution.

**Comment by xomx** [comment]
> I rechecked again. 7.5.1 is working, 7.5.2 not.

So the culprit there was the change in toolset, as we have already once discovered (MS crippled somehow the CRT _stat functionality):
https://community.notepad-plus-plus.org/topic/15089/notepad-file-status-auto-detection-not-working/19

But the relevant N++ code has since been rewritten, now it uses the WINAPI GetFileAttributesEx instead.

**Comment by xomx** [comment]
@LeMerP 
Could you please elaborate, what type of file system is used on your NSS volume?
And could you check yourself (e.g. from the system shell) the timestamp of the file being reported in the nppLogNetworkDriveIssue.log?

**Comment by xomx** [comment]
@donho 
It might be worth to try taking the current v8.2 source code and prepare from it a configurable testing version, where the `GetFileAttributesEx` way can be changed back to the `stat` POSIX one. MS probably has fixed its `_stat` problem in higher platform toolsets. If that solution will work, we will have a optional setting for all the problematic *nix networks storage resources.

**Comment by donho** [comment]
@xomx 
I have no idea how many users of Notepad++ world-wide, but I know there is a huge amount of Notepad++ users - we cannot allow to change the basic functions/API like that for a special isolated case.

**Comment by xomx** [comment]
@donho 
> we cannot allow to change the basic functions/API like that for a special isolated case

I agree. I did not mean changing the existing GetFileAttributesEx way for all the N++ users.

I mean writing e.g. GetFileAttributesExNpp substitute like:
```
BOOL GetFileAttributesExNpp(LPCTSTR lpFileName, GET_FILEEX_INFO_LEVELS fInfoLevelId, LPVOID lpFileInformation)
{
	if (!_bPosixGetFileAttributesEx)
		return ::GetFileAttributesEx(lpFileName, fInfoLevelId, &lpFileInformation);

	BOOL bRet = false;
	
	// here substitute all the GetFileAttributesEx functionality with the POSIX _stat
	...

	return bRet;
}
```

**Comment by LeMerP** [comment]
> @LeMerP Could you please elaborate, what type of file system is used on your NSS volume? (Edited: Ok, the NSS is the Novell [NSS](https://en.wikipedia.org/wiki/Novell_Storage_Services) filesystem, I did not know that.)
> 
> And could you check yourself (e.g. from the Microfocus system shell) the timestamp of the file being reported in the nppLogNetworkDriveIssue.log?
> 
> Edited: If the NFS file access protocol is used(?), timestamps can be unreliable (a delayed write caching or a time de-synchronization between client and server in action).

@xomx 
Sorry for the delay. Did some testing. What I can say:
"nssvol" is some kind of modified ext(2) filesystem.
The issue has to be related to the used Novell Core Protocol (NCP). If I am accessing the NSS volume with CIFS protocol from Windows, the issue does not occur. 
nppLogNetworkDriveIssue.log stats 08:41:21 as modified file time whereas at the server it is 08:41:15.
So it might be some file caching thing that is special for NCP.

**Comment by xomx** [comment]
@LeMerP 
> some kind of modified ext(2) filesystem

EXT2/3 timestamp granularity is probably 1s only (default inode size there is 128 B in comparison e.g. to EXT4, which has 256 B and therefore nanosecond timestamps precision) 

> with CIFS protocol from Windows, the issue does not occur

Good. So can we consider this issue as solved?
(even if you close this issue now, you can add later some new stuff in it when necessary and we will receive a notification about it)

> nppLogNetworkDriveIssue.log stats 08:41:21 as modified file time whereas at the server it is 08:41:15

Sorry, my mistake, I did not express myself clearly. What I mean:
- reproduce the error state once more
- after the issue manifests again, post here the content of the nppLogNetworkDriveIssue.log again
- then check the real text file timestamp on the NSS volume (`\\oesserver\nss\test.txt`) by the help of the system there and post this timestamp too

My interest - one of the previously reported timestamp by you has its millisecond part set to zero:
  `attributes.ftLastWriteTime (2022/01/7 08:21:28.0) < _timeStamp (2022/01/7 08:21:28.689)`
This is weird, because of it means, that the actual file timestamp is a little bit older than the N++ timestamp stored.
The problem could be the:
- caching
- de-synchronization between client and server in time
- the real timestamp granularity at the final destination is 1s

**Comment by LeMerP** [comment]
Hey guys,

sorry for delay, again.
If think I found the root cause of this issue: in the past we had to enable a special function of "OES Client" that is called "Lazy Close".  Here is a short description of this "feature":
When File Caching is enabled, delays the File Close on the network allowing the file to be reopened without accessing the network.

Btw, "File Caching" is also enabled.

If I am disabling this lazy close option Notepad++ will not display the "modified by another program" dialog.

PS: we found another application that has a similar issue like Notepad++: XMind (Version: XMind 8 Update 9 (R3.7.9.201912052356)

I think Microfocus has to deal with this issue.
